mqtt_jmeter/src/main/java/net/xmeter/samplers/SubSampler.java

```diff
@@ -155,19 +163,33 @@ public SampleResult sample(Entry arg0) {
 			}
 		} else {
 			synchronized (dataLock) {
-				int receivedCount1 = (batches.isEmpty() ? 0 : batches.element().getReceivedCount());;
-				boolean needWait = false;
-				if(receivedCount1 < sampleCount) {
-					needWait = true;
-				}
-				
-				if(needWait) {
+				int receivedCount = (batches.isEmpty() ? 0 : batches.element().getReceivedCount());
+				if (receivedCount < sampleCount) {
 					try {
-						dataLock.wait();
+						if (sampleCountTimeout > 0) {
+							// handle spurious wakeups (https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Object.html#wait(long,int))
+							lockReleased = false;
+							long endtime = System.currentTimeMillis() + sampleCountTimeout;
+							long currenttime = 0;
+							while (!lockReleased || currenttime < endtime) {
```

Comment:
This should be `&&` otherwise timeout will never occur

Reply:
right!